### PR TITLE
Settings page, share intent and location-based notification

### LINKED ISSUE #90: As an app user, I want to be notified when a new favor is posted in the area and check the details once I click on the notification

### LINKED ISSUE #91: As a user, I want to have a page where I can see/change the current settings of the app (e.g. suscribe for notifications, radius of notifications etc.)

### LINKED ISSUE #94: Navigation menu should not open up when swiping right on the screen

### PR DESCRIPTION
**Pull request by mikanikos**
Fix #94, closes #90, closes #91

### CODE REVIEW

File: app/src/test/java/ch/epfl/favo/util/LocationManagerDependencyFacotryTest.java [hunk 1]
```diff
@@ -4,36 +4,38 @@
 import android.content.Intent;
 import android.location.LocationManager;
```
Comment (danielpgp1012) [inline app/src/test/java/ch/epfl/favo/util/LocationManagerDependencyFacotryTest.java]: there's a typo in the class name

Comment (mikanikos) [reply app/src/test/java/ch/epfl/favo/util/LocationManagerDependencyFacotryTest.java]: You're right, fixed in the next commit.

File: app/src/main/java/ch/epfl/favo/user/User.java [hunk 2]
```diff
@@ -64,34 +64,38 @@ public String getNotificationId() {
...
+  }
+
   public LocalDate getBirthDate() {
     return birthDate;
   }
 
-  public Location getLocation() { return location; }
+  public Location getLocation() {
+    return location;
+  }
```
Comment (muneebazam) [inline app/src/main/java/ch/epfl/favo/user/User.java]: any reason why you moved this function defintion up? not a big deal but was wondering because we had getters and setters both listed in order of fields

File: server/functions/index.js [hunk 3]
```diff
@@ -3,31 +3,73 @@ const functions = require('firebase-functions');
...
+function distanceInKm(lon1, lat1, lon2, lat2) {
+    var R = 6371; // Radius of the earth in km
+    var dLat = (lat2 - lat1) * Math.PI / 180;  // Javascript functions in radians
+    var dLon = (lon2 - lon1) * Math.PI / 180;
+    var a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
+        Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) *
+        Math.sin(dLon / 2) * Math.sin(dLon / 2);
+    var c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
+    return R * c;
+}
```
Comment (arosu) [inline server/functions/index.js]: Maybe we can find libraries that do this for us? Just a thought.

File: app/src/main/java/ch/epfl/favo/util/FakeFavorList.java [hunk 4]
```diff
@@ -7,37 +7,41 @@
 import ch.epfl.favo.favor.Favor;
```
Comment (muneebazam) [inline app/src/main/java/ch/epfl/favo/util/FakeFavorList.java]: what is the  purpose of this file, where is it being used? and if it is  only being used for tests i think its better to move this into the test directory

File: server/functions/index.js [hunk 5]
```diff
@@ -3,31 +3,73 @@ const functions = require('firebase-functions');
...
 
-        // This registration token comes from the client FCM SDKs.
-        var registrationToken = 'TOKEN HERE';
+function sendMessage(message, usersIds) {
+    admin.messaging().sendMulticast(message)
+        .then((response) => {
+            if (response.failureCount > 0) {
+                const failedTokens = [];
```
Comment (arosu) [inline server/functions/index.js]: Curious as to why this is a const when you modify by pushing on line 27.

File: app/src/main/java/ch/epfl/favo/view/tabs/MapsPage.java [hunk 6]
```diff
@@ -54,34 +51,45 @@
...
   @Override
   public void onMapReady(GoogleMap googleMap) {
     setupView();
     mMap = googleMap;
     mMap.clear();
     drawSelfLocationMarker();
     drawFavorMarker(updateFavorlist());
-    //throw new RuntimeException(
-      //  "This exception will " + "certainly be thrown out as long as Map is ready");
+    // throw new RuntimeException(
```
Comment (muneebazam) [inline app/src/main/java/ch/epfl/favo/view/tabs/MapsPage.java]: is this line throwing exception still needed

File: server/functions/index.js [hunk 7]
```diff
@@ -3,31 +3,73 @@ const functions = require('firebase-functions');
...
+                });
+                console.log('List of tokens that caused failures: ' + failedTokens);
+            }
+        });
+}
 
-        var message = {
+// send new favor notification to users in the area
+exports.sendNotificationNearbyOnNewFavor = functions.firestore
+    .document('/favors/{favorId}')
```
Comment (arosu) [inline server/functions/index.js]: I'd write the following so it's a bit more readable.
```
firestore
  .collection("favors")
  .document(favorId)
```

File: server/firebase.json [hunk 8]
```diff
@@ -1 +1,10 @@
-{}
+{
```
Comment (muneebazam) [inline server/firebase.json]: what is this used for?

File: server/functions/index.js [hunk 9]
```diff
@@ -3,31 +3,73 @@ const functions = require('firebase-functions');
...
+    .onCreate((change) => {
+
+        // get new favor that has just been posted
+        const newFavor = change.data();
+        var latFav = newFavor.location.latitude;
+        var longFav = newFavor.location.longitude;
+        var usersIds = [];
+        var maxDistance = 100.0; // in km
```
Comment (arosu) [inline server/functions/index.js]:
```suggestion
        private static final Float MAX_DISTANCE = 100.0; // in km
```

File: app/src/androidTest/java/ch/epfl/favo/FavorAndroidTest.java [hunk 10]
```diff
@@ -10,54 +10,51 @@
...
+  public GrantPermissionRule permissionRule =
+      GrantPermissionRule.grant(android.Manifest.permission.ACCESS_FINE_LOCATION);
+
+  @After
+  public void tearDown() {
+    DependencyFactory.setCurrentFirebaseUser(null);
+  }
+
+  @Test
+  public void favorGettersReturnCorrectValuesByParcelable() {
```
Comment (muneebazam) [inline app/src/androidTest/java/ch/epfl/favo/FavorAndroidTest.java]: this function does not actually test anything, are the asserts supposed to be commented out?

File: server/functions/index.js (hunk 9 above)
Comment (arosu) [reply server/functions/index.js]: Nevermind, I realized this is javascript lol

File: app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java [hunk 11]
```diff
@@ -14,28 +14,29 @@
...
-        assert favor.getDescription() != null;
-        favorTitle.setText(favor.getTitle());
-        favorDesc.setText(favor.getDescription());
-        // Return the completed view to render on screen
-        return convertView;
+  @Override
+  public View getView(int position, View convertView, ViewGroup parent) {
+    // Get the data item for this position
+    Favor favor = getItem(position);
+    assert favor != null;
```
Comment (arosu) [inline app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java]: I would do something like
```
Favor favor;
try {
  favor = getItem(position);
} catch (Exception e) {
  throw e;
}
```
(but maybe figure out exactly what types of exceptions `getItem` can throw.

File: server/functions/index.js (hunk 7 above)
Comment (mikanikos) [reply server/functions/index.js]: I'm not really sure it would work the same way, {favorId} is just a placeholder for indicating any favor in the favors collection

File: app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java [hunk 12]
```diff
@@ -14,28 +14,29 @@
...
+    if (convertView == null) {
+      convertView =
+          LayoutInflater.from(getContext()).inflate(R.layout.favor_list_item, parent, false);
     }
+    // Lookup view for data population
+    TextView favorTitle = convertView.findViewById(R.id.title);
+    TextView favorDesc = convertView.findViewById(R.id.desc);
+    // Populate the data into the template view using the data object
+    assert favor.getTitle() != null;
+    assert favor.getDescription() != null;
```
Comment (arosu) [inline app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java]: Same thing as above here. In general, we want to avoid using assertions in code in the main package and save it for tests. Try/catch blocks would be more appropriate.

File: server/functions/index.js (hunk 5 above)
Comment (mikanikos) [reply server/functions/index.js]: You're probably right, I'm not really into Javascript but I remember this was an example from the Firebase documentation so we can just keep it like this for the moment.

File: app/src/main/java/ch/epfl/favo/view/tabs/addFavor/FavorRequestView.java [hunk 13]
```diff
@@ -17,18 +17,16 @@
...
- */
-public class FavorRequestView extends BottomDestinationTab implements View.OnClickListener {
-
+import ch.epfl.favo.view.ViewController;
 
+/** allows user to create favor objects that can be posted in the DB */
+public class FavorRequestView extends Fragment implements View.OnClickListener {
+  // TODO: Rename parameter arguments, choose names that match
+  // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
+  private static final String ARG_PARAM1 = "param1";
```
Comment (muneebazam) [inline app/src/main/java/ch/epfl/favo/view/tabs/addFavor/FavorRequestView.java]: i think we can remove these two variables since they arent used, if we have to keep them we should consider renaming them  and removing the comment above it

File: server/functions/index.js (hunk 3 above)
Comment (mikanikos) [reply server/functions/index.js]: Yeah I looked for it but my inexperience with Javascript just made me do this with a manual approach. We should definitely look for some library to handle this in the next sprints. We don't have enough time right now to test another library and check that everything works the same but I'll open an issue for this.

File: app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java [hunk 14]
```diff
@@ -14,28 +14,29 @@
...
+    assert favor != null;
+    // Check if an existing view is being reused, otherwise inflate the view
+    if (convertView == null) {
+      convertView =
+          LayoutInflater.from(getContext()).inflate(R.layout.favor_list_item, parent, false);
     }
+    // Lookup view for data population
+    TextView favorTitle = convertView.findViewById(R.id.title);
+    TextView favorDesc = convertView.findViewById(R.id.desc);
+    // Populate the data into the template view using the data object
```
Comment (muneebazam) [inline app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java]: we should be using @Nullable or Try/Catches for in code. Assert should normally only ever be used for testing because in the case that any one of  this is null its going to throw an assertionFailed and the code will not be able to recover/crash

Comment (muneebazam) [reply app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java]: it would be best to use a try/catch with defaults in case anything is null because otherwise it will crash

File: app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java (hunk 11 above)
Comment (mikanikos) [reply app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java]: I actually didn't modify this part, it's probably in the changed files because I reformatted and cleaned up automatically the whole project but I agree with the suggestion

File: app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java (hunk 12 above)
Comment (mikanikos) [reply app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java]: Yeah I agree, this is still part of an old merge in the master that I didn't touch apart from reformatting

File: app/src/main/java/ch/epfl/favo/common/DatabaseWrapper.java [hunk 15]
```diff
@@ -2,58 +2,60 @@
...
-            firestore.setFirestoreSettings(settings);
-        } catch (Exception e) {
-            Log.d(TAG, e.toString());
-        }
+  private static final String TAG = "DatabaseWrapper";
+  private static final int ID_LENGTH = 25;
+  private static DatabaseWrapper INSTANCE = null;
+  // final fields regarding ID generation
+  private static String ID_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+  private FirebaseFirestore firestore;
```
Comment (arosu) [inline app/src/main/java/ch/epfl/favo/common/DatabaseWrapper.java]: This comment is now in the wrong place since `ID_LENGTH` is now above the comment.

File: app/src/main/java/ch/epfl/favo/common/DatabaseWrapper.java [hunk 16]
```diff
@@ -2,58 +2,60 @@
...
-        return INSTANCE;
+  public static DatabaseWrapper getInstance() {
+    if (INSTANCE == null) {
+      INSTANCE = new DatabaseWrapper();
     }
+    return INSTANCE;
+  }
 
-    public static String generateRandomId() {
+  public static String generateRandomId() {
```
Comment (arosu) [inline app/src/main/java/ch/epfl/favo/common/DatabaseWrapper.java]: Not relevant to this PR, but I'm assuming there was trouble with Java's Random package so we're doing the generation ourselves now right @muneebazam?

Comment (muneebazam) [reply app/src/main/java/ch/epfl/favo/common/DatabaseWrapper.java]: yeah, the ID_CHARS is ugly but best we can do without using any external library

File: app/src/main/java/ch/epfl/favo/auth/SignInActivity.java [hunk 17]
```diff
@@ -48,7 +48,7 @@ protected void onCreate(Bundle savedInstanceState) {
     startActivityForResult(createSignInIntent(), RC_SIGN_IN);
   }
 
-  private void checkPlayServices () {
+  private void checkPlayServices() {
```
Comment (arosu) [inline app/src/main/java/ch/epfl/favo/auth/SignInActivity.java]: Just as a suggestion, try not to include changes like this in a PR if this file is not relevant to the PR. A better solution would be to make another PR and call it something like "Cleaning up some code". It would make the lives of people reviewing this PR a lot easier.

File: app/src/main/java/ch/epfl/favo/common/DatabaseWrapper.java (hunk 15 above)
Comment (muneebazam) [reply app/src/main/java/ch/epfl/favo/common/DatabaseWrapper.java]: yeah the fields were correctly sorted before, can we please move them back into that order

File: app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java (hunk 12 above)
Comment (arosu) [reply app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java]: Ahh okay I didn't notice that, Git was telling me these lines were added so I thought it was part of the PR.

File: app/src/androidTest/java/ch/epfl/favo/notifications/FirebaseMessagingServiceTest.java [hunk 18]
```diff
@@ -29,52 +29,55 @@
...
+    bundle.putString("gcm.notification.title", NOTIFICATION_TITLE);
+    bundle.putString("gcm.notification.body", NOTIFICATION_BODY);
+    bundle.putString("gcm.notification.e", "1");
+
+    FirebaseMessagingService.showNotification(
+        mainActivityTestRule.getActivity(),
+        Objects.requireNonNull(new RemoteMessage(bundle).getNotification()),
+        "Default channel id");
+
+    // WORKS LOCALLY, NOT ON TRAVIS
```
[the remark this answers is not in the document]
Comment (mikanikos) [reply app/src/androidTest/java/ch/epfl/favo/notifications/FirebaseMessagingServiceTest.java]: Well, I still hope to find a solution to make that part of the test work on Travis but I guess we can remove it for now.

[the remark this answers is not in the document]
Comment (muneebazam) [reply app/src/androidTest/java/ch/epfl/favo/notifications/FirebaseMessagingServiceTest.java]: ok ya sounds good, what I do  when i have stuff i need for later is to git stash or git temp branch it that way we dont push unnecessary things into master

File: server/functions/index.js [hunk 19]
```diff
@@ -3,31 +3,74 @@ const functions = require('firebase-functions');
...
+                console.log('List of tokens that caused failures: ' + failedTokens);
+            }
+        });
+}
+
+// send new favor notification to users in the area
+exports.sendNotificationNearbyOnNewFavor = functions.firestore
     .document('favors/{favorId}')
-    .onWrite((change, context) => {
+    .onCreate((change) => {
```
[the remark this answers is not in the document]
Comment (arosu) [reply server/functions/index.js]: Fuck off codeclimate.

File: server/functions/index.js [hunk 20]
```diff
@@ -3,31 +3,75 @@ const functions = require('firebase-functions');
 
 const admin = require('firebase-admin');
 admin.initializeApp();
+var db = admin.firestore();
 
-exports.sendNotificationOnDBChange = functions.firestore
+exports.sendNotificationNearbyOnNewFavor = functions.firestore
     .document('favors/{favorId}')
-    .onWrite((change, context) => {
+    .onCreate((change, context) => {
```
[the remark this answers is not in the document]
Comment (arosu) [reply server/functions/index.js]: No one likes you codeclimate.

File: server/functions/index.js (hunk 3 above)
Comment (arosu) [reply server/functions/index.js]: Sounds good, thanks!

File: server/functions/index.js (hunk 5 above)
Comment (arosu) [reply server/functions/index.js]: I'm actually curious as to why we have to use JavaScript in general?

File: app/src/main/java/ch/epfl/favo/MainActivity.java [hunk 21]
```diff
@@ -30,36 +31,26 @@
  */
 public class MainActivity extends AppCompatActivity
     implements NavigationView.OnNavigationItemSelectedListener, ViewController {
+  private static final String TAG = "MainActivity";
+  // Bottom tabs
+  public RadioButton mapButton;
+  public RadioButton favListButton;
   // UI
-  //private AppBarConfiguration appBarConfiguration;
+  // private AppBarConfiguration appBarConfiguration;
```
[the remark this answers is not in the document]
Comment (mikanikos) [reply app/src/main/java/ch/epfl/favo/MainActivity.java]: Agree

File: server/functions/index.js (hunk 7 above)
Comment (arosu) [reply server/functions/index.js]: Yeah you're probably right. I wrote this comment before I realized this was javascript code lol.

File: app/src/main/java/ch/epfl/favo/user/User.java (hunk 2 above)
Comment (mikanikos) [reply app/src/main/java/ch/epfl/favo/user/User.java]: This was done automatically by the reformat tool of Android studio, I'll change it back as it was before

File: app/src/androidTest/java/ch/epfl/favo/notifications/FirebaseMessagingServiceTest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (arosu) [reply app/src/androidTest/java/ch/epfl/favo/notifications/FirebaseMessagingServiceTest.java]: Or maybe write it as a TODO.
`TODO (mikanikos): get this working on Travis!`

File: app/src/main/java/ch/epfl/favo/user/User.java (hunk 2 above)
Comment (muneebazam) [reply app/src/main/java/ch/epfl/favo/user/User.java]: oh i must;ve forgot to reformat this one when i added it, sorry about that

Comment (arosu) [reply app/src/main/java/ch/epfl/favo/user/User.java]: This is the way google format does it.

File: app/src/main/java/ch/epfl/favo/util/FakeFavorList.java (hunk 4 above)
Comment (mikanikos) [reply app/src/main/java/ch/epfl/favo/util/FakeFavorList.java]: Another closed PR had this file, we should ask the author of this file for the purpose. I'll keep it there because I don't want to break things.

File: app/src/main/java/ch/epfl/favo/view/tabs/MapsPage.java (hunk 6 above)
Comment (mikanikos) [reply app/src/main/java/ch/epfl/favo/view/tabs/MapsPage.java]: Agree, I'll remove it

File: server/firebase.json (hunk 8 above)
Comment (mikanikos) [reply server/firebase.json]: Agree I'll remove it

Comment (mikanikos) [reply server/firebase.json]: Agree, I'll remove it

File: app/src/androidTest/java/ch/epfl/favo/FavorAndroidTest.java (hunk 10 above)
Comment (mikanikos) [reply app/src/androidTest/java/ch/epfl/favo/FavorAndroidTest.java]: Since this is done by someone else, I prefer not to touch his code because it might be relevant for him

File: app/src/main/java/ch/epfl/favo/view/tabs/addFavor/FavorRequestView.java (hunk 13 above)
Comment (mikanikos) [reply app/src/main/java/ch/epfl/favo/view/tabs/addFavor/FavorRequestView.java]: Agree, I'll remove it even though it's not my code

Comment (mikanikos) [reply app/src/main/java/ch/epfl/favo/view/tabs/addFavor/FavorRequestView.java]: Sorry changed my mind, I prefer to remove only things that are really not necessary and don't affect others. Since this is by someone else, I'll not touch it

**Comment by muneebazam** [comment]
you can just review the comments i left which are related to files that you have changed, sorry for all the confusion. Going forward lets only reformat files that each person works on to keep things simple.

The code itself looks good to me so I have approved the change

File: app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java (hunk 14 above)
Comment (mikanikos) [reply app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java]: Agree, see above for the answer I gave Alex

File: app/src/main/java/ch/epfl/favo/common/DatabaseWrapper.java (hunk 15 above)
Comment (mikanikos) [reply app/src/main/java/ch/epfl/favo/common/DatabaseWrapper.java]: Changed as it was before

File: app/src/main/java/ch/epfl/favo/auth/SignInActivity.java (hunk 17 above)
Comment (mikanikos) [reply app/src/main/java/ch/epfl/favo/auth/SignInActivity.java]: You're right but this should not happen if everyone formats his own code before pushing. Agree we should make another PR for cleaning and reformatting.

File: app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java (hunk 11 above)
Comment (mikanikos) [reply app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java]: We should ask the author of the file for the reasons of this and to change it properly.

File: app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java (hunk 12 above)
Comment (mikanikos) [reply app/src/main/java/ch/epfl/favo/view/tabs/favorList/FavorAdapter.java]: We should ask the author of the file for the reasons of this and to change it properly.

File: server/firebase.json (hunk 8 above)
Comment (mikanikos) [reply server/firebase.json]: Btw, it just contains config for the firebase nodejs environment (like emulators)

**Comment by mikanikos** [comment]
Unfortunately this PR was huge because of the reformatting done by Android studio and there were lots of dirty, bad-formatted files in the master. As you guys said, next time let's try only reformat files that each of us worked on and keep our code as clean as possible so that these kind of problems don't happen again. We should also keep in mind to review all the PRs done by everyone at least quickly and comment on the actual files that were modified by the creators and ask them for reasons because I'm in no position to change things that are already approved in the master and I don't know what they are.

**Comment by muneebazam** [comment]
Agree entirely, let’s talk about this next team meeting so we can make it a priority for next sprint!
